grpc/src/grpc.rs

```diff
@@ -36,6 +42,177 @@ pub use crate::grpc::cosmos_tx::JsBroadcastMode;
+    pub(crate) fn append_metadata_map(&mut self, metadata: &MetadataMap) {
+        for key_and_value in metadata.iter() {
+            match key_and_value {
+                KeyAndValueRef::Ascii(key, val) => {
+                    self.metadata.append(key, val.clone());
+                }
+                KeyAndValueRef::Binary(key, val) => {
+                    self.metadata.append_bin(key, val.clone());
+                }
+            }
+        }
+    }
+
+    /// Merges the other context into self.
+    pub(crate) fn extend(&mut self, other: &Context) {
+        self.append_metadata_map(&other.metadata);
+    }
+}
+
+/// A call of the grpc method.
+///
+/// Allows setting additional context for request before awaiting
+/// the call.
+///
+/// ```
+/// # use celestia_grpc::{Result, GrpcClient};
+/// # use celestia_grpc::grpc::TxPriority;
+/// # use celestia_types::state::Address;
+/// # async |client: GrpcClient, addr: &Address| -> Result<()> {
+/// let balance = client.get_balance(addr, "utia")
+///     .metadata("x-token", "your secret token")?
+///     .block_height(12345)
+///     .await?;
+/// # Ok(())
+/// # };
+/// ```
+pub struct AsyncGrpcCall<Response, Error = crate::Error> {
+    call: CallFn<Response, Error>,
+    pub(crate) context: Context,
+}
+
+impl<Response, Error> AsyncGrpcCall<Response, Error> {
+    /// Create a new grpc call out of the given function.
+    ///
+    /// This method is exposed only for internal use in `celestia-client`.
+    /// It is not considered a public API and thus have no SemVer guarantees.
+    #[doc(hidden)]
+    pub fn new<F, Fut>(call_fn: F) -> Self
+    where
+        F: FnOnce(Context) -> Fut + 'static,
+        Fut: Future<Output = Result<Response, Error>> + Send + 'static,
+    {
+        Self {
+            call: Box::new(|context| call_fn(context).boxed()),
+            context: Context::default(),
+        }
+    }
+
+    /// Extend the current context of the grpc call with provided context
+    ///
+    /// This method is exposed only for internal use in `celestia-client`.
+    /// It is not considered a public API and thus have no SemVer guarantees.
+    #[doc(hidden)]
+    pub fn context(mut self, context: &Context) -> Self {
+        self.context.extend(context);
+        self
+    }
+
+    /// Append an ascii metadata to the grpc request.
+    pub fn metadata(mut self, key: &str, val: &str) -> Result<Self, MetadataError> {
+        self.context.append_metadata(key, val)?;
+        Ok(self)
+    }
+
+    /// Append a binary metadata to the grpc request.
+    ///
+    /// Keys for binary metadata must have `-bin` suffix.
+    pub fn metadata_bin(mut self, key: &str, val: &[u8]) -> Result<Self, MetadataError> {
+        self.context.append_metadata_bin(key, val)?;
+        Ok(self)
+    }
+
+    /// Append a metadata map to the grpc request.
+    pub fn metadata_map(self, metadata: MetadataMap) -> Self {
+        let context = Context { metadata };
+        self.context(&context)
```

Comment:
I think calling `self.context.append_metadata_map` is more correct. We may have more stuff in `Context` later on.